tools-in-rust-repo.md

```diff
@@ -0,0 +1,139 @@
+# Maintaining tools in the Rust repository
+
+This document contains information for tool maintainers and Rust contributors
+who need to work with tools that are part of the Rust distribution and
+repository. For information on requirements for such tools see the
+[stability docs](https://github.com/nrc/dev-tools-team/blob/master/stability/README.md).
+
+When adding tools or making significant changes, please be sure to cc @nrc and
+@aclexcrichton on PRs. Adding tools to the Rust repo requires approval from the
+dev-tools team and sign-off from the core team.
+
+
+## Repository layout
+
+Tools should live in their own repository which should be in an official
+organisation (e.g., [rust-lang-nursery](https://github.com/rust-lang-nursery)).
+Tools should be included into the Rust repository as a Git submodule in the
+[src/tools](https://github.com/rust-lang/rust/tree/master/src/tools) directory.
+
+
+## Constraints on tools
+
+It is essential that any commit used as a submodule in the Rust repo is
+preserved forever (i.e., is *never* deleted or changed). This ensures that we
+will continue to be able to build any historic release of Rust (which is
+important for bisecting regressions, etc). In order to do this, any commit
+pulled into the Rust repo as a submodule must be from a Rust organisation repo,
+not an author's. If the commit is only on a branch, that branch cannot be
+deleted. Whether the commit is on a branch or master, it cannot be rebased
+(which essentially deletes the commit and creates a new one).
+
+Tools used in the Rust repo cannot use Cargo workspaces. This is because the
+tool itself will be treated as a workspace crate and workspaces cannot be
+nested.
+
+
+## Making breaking changes
+
+If making a change to a 'core' part of Rust causes breakage in any of the tools,
+then there is a bit of a dance to perform to land the Rust PR.
+
+Communication is key here! If you're a compiler dev, communicate with tool devs
+as soon as possible (and vice versa).
+
+At the end of the day, the tool's maintainers are responsible for ensuring that
+their tool continues to build and pass tests in the Rust repo. For non-trivial
+fixes, tool devs should be happy to help fix breaking changes.
+
+How to land a breaking change:
```

Comment:
Sorry, I didn't see this comment before merging.

The tool CI never breaks on master, only on a branch. So, I think it depends on the breaking change - so far with the RLS, the changes have been trivial to fix (e.g., adding `()` to every use of `span.lo/hi`) and the compiler changer has made the change, and I've done the branch dance on the RLS-side (e.g., https://github.com/rust-lang/rust/pull/43968). I imagine that if the change needed more work in the tool, then the tool authors would do the work, but it would be basically the same process (I've tried to stress that it is the tool authors responsibility here).

The trouble with the exceptions route is that it leads to a tool not being built, which means that rustup will look for it and not find it, and presumably error out. We really need infrastructure support before I'm happy recommending that route. (And in practice the non-exception route is working OK, and with the incremental path to enabling tools, we should have time to observe if it is continuing to work OK).